File: test/strategy/bench_kelly.py
#!/usr/bin/env python3

import time
import datetime
import eatmud


def bench_kelly(trans: eatmud.Transaction, start_date: eatmud.Date
                ) -> list[list[float]]:
    ns = [1300, 1600]
    inflations = [0.015, 0.015]
    risk_bounds = [0.01, 0.01]
    results = []
    for save_log in (True, False,):
        for save_record in (True, False):
            name = f'save_log={save_log}, save_record={save_record}'
            now = time.time()
            res = []
            for weekday in range(0, 5):
                it = trans.iter(save_log, save_record)
                it.goto(start_date)
                it.inflow(1.)
                eatmud.strategy.kelly_weekly(it, eatmud.Weekday(
                    weekday), ns, inflations, risk_bounds)
                res.append(it.asset())
            total_time = time.time() - now
            print(f'running kelly({name}) took '
                  f'{total_time*1000} milli seconds')
            results.append(res)
    return results


def test_kelly():
    hs300 = eatmud.Fund.from_stock(eatmud.io.read_tdx("tdx/test-hs300.txt"))
    gz2000 = eatmud.Fund.from_stock(eatmud.io.read_tdx("tdx/test-gz2000.txt"))
    start_date = datetime.datetime.strptime('20170101', '%Y%m%d').date()
    end_date = datetime.datetime.strptime('20240101', '%Y%m%d').date()
    trans = eatmud.Transaction([hs300, gz2000], None, end_date)
    results = bench_kelly(trans, start_date)
    print(results)


if __name__ == '__main__':
    test_kelly()

File: eatmud_core/src/bin/bench_kelly.rs
use std::time::Instant;

use eatmud::{Fund, NaiveDate, Transaction, Weekday, io::read_tdx};

fn bench_kelly(trans: &Transaction, start_date: NaiveDate) -> Vec<Vec<f64>> {
    let ns = [1300, 1600];
    let inflations = [0.015, 0.015];
    let risk_bounds = [0.01, 0.01];
    let mut results = Vec::new();
    for save_log in [true, false] {
        for save_record in [true, false] {
            let name = format!("save_log={}, save_record={}", save_log, save_record);
            let now = Instant::now();
            let mut res = Vec::new();
            for weekday in 0..5 {
                let mut it = trans.iter(save_log, save_record);
                it.goto(start_date);
                it.inflow(1.).unwrap();
                eatmud::strategy::kelly_weekly(
                    &mut it,
                    Weekday::try_from(weekday).unwrap(),
                    &ns,
                    &inflations,
                    &risk_bounds,
                )
                .unwrap();
                res.push(it.asset());
            }
            let total_time = Instant::now() - now;
            println!(
                "running kelly({}) took {} milli seconds",
                name,
                total_time.as_micros() as f64 / 1000.
            );
            results.push(res);
        }
    }
    results
}

fn main() {
    let hs300 = Fund::from(&read_tdx("tdx/test-hs300.txt").unwrap());
    let gz2000 = Fund::from(&read_tdx("tdx/test-gz2000.txt").unwrap());
    let start_date = NaiveDate::parse_from_str("20170101", "%Y%m%d").unwrap();
    let end_date = NaiveDate::parse_from_str("20240101", "%Y%m%d").unwrap();
    let trans = Transaction::new(&[&hs300, &gz2000], None, Some(end_date));
    let results = bench_kelly(&trans, start_date);
    println!("{:?}", results);
    assert_eq!(results[0], results[1]);
    assert_eq!(results[0], results[2]);
    assert_eq!(results[0], results[3]);
}
